### PR TITLE
Add snappy compression between client and regionserver

### PR DESCRIPTION
**Pull request by timoha**
- Implements cellblocks serialization for mutate and multi
- Adds snappy compression and decompression for cellblocks

### CODE REVIEW

File: region/client.go [hunk 1]
```diff
@@ -621,11 +642,20 @@ func (c *client) send(rpc hrpc.Call) (uint32, error) {
...
 
-	if err := c.write(b); err != nil {
+	if cellblocks != nil {
+		bfs := append(net.Buffers{b}, cellblocks...)
+		_, err = bfs.WriteTo(c.conn)
+	} else {
+		err = c.write(b)
+	}
+	// free header/request buffer
+	freeBuffer(b)
```
Comment (tsuna) [inline region/client.go:678]: This used to be done via a `defer`, shouldn't it still be?

File: region/client_test.go [hunk 2]
```diff
@@ -942,6 +958,131 @@ func TestSanity(t *testing.T) {
...
+		conn:          mockConn,
+		rpcs:          make(chan hrpc.Call),
+		done:          make(chan struct{}),
+		sent:          make(map[uint32]hrpc.Call),
+		rpcQueueSize:  1, // size one to skip sendBatch
+		flushInterval: 1000 * time.Second,
+		compressor:    &compressor{Codec: mockCodec{}},
+	}
+
+	var wg sync.WaitGroup
```
Comment (tsuna) [inline region/client_test.go:975]: Nit: Only ever `Add()`ed to once, you could just use channel that gets closed.

File: region/compressor.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+	"github.com/tsuna/gohbase/compression"
+)
+
+type compressor struct {
+	compression.Codec
+}
+
+func min(x, y uint32) int {
+	if x > y {
```
Comment (tsuna) [inline region/compressor.go:21]: that's cute, implementing `min()` with a greater than check :)

File: region/compressor.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		b, chunkLen = c.Encode(uncompressedBuffer[:n], b)
+
+		// write the chunk length
+		binary.BigEndian.PutUint32(b[lenOffset:], chunkLen)
+
+		if err != nil {
+			break
+		}
+	}
+	freeBuffer(uncompressedBuffer)
```
Comment (tsuna) [inline region/compressor.go:63]: shouldn't this be called via `defer`?

File: region/compressor.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+		// grow for chunk length
+		lenOffset = len(b)
+		b = growBuffer(b, 4)
+
+		b, chunkLen = c.Encode(uncompressedBuffer[:n], b)
+
+		// write the chunk length
+		binary.BigEndian.PutUint32(b[lenOffset:], chunkLen)
+
+		if err != nil {
```
Comment (tsuna) [inline region/compressor.go:59]: is any error expected here other than EOF?

File: region/compressor.go (hunk 3 above)
Comment (timoha) [reply region/compressor.go]: Brain works in mysterious ways :D

File: region/client_test.go (hunk 2 above)
Comment (timoha) [reply region/client_test.go:975]: used on line 1046 as well

File: region/compressor.go (hunk 4 above)
Comment (timoha) [reply region/compressor.go]: Done

File: region/compressor.go (hunk 5 above)
Comment (timoha) [reply region/compressor.go]: Done, panicing on any other error

**Pull request merged**